### PR TITLE
Read interface statistics from operating system via pnal

### PR DESCRIPTION
**Pull request by pyhys**
Adapt pnal ethernet-related functions to use the interface name as argument.

### CODE REVIEW

File: src/ports/linux/pnal.c [hunk 1]
```diff
@@ -256,6 +257,187 @@ int pnal_set_ip_suite (
...
+   eth_status_linux.cmd = ETHTOOL_GSET;
+
+   if (ioctl (control_socket, SIOCETHTOOL, &ifr) >= 0)
+   {
+      speed = ethtool_cmd_speed (&eth_status_linux);
+
+      status->is_autonegotiation_enabled =
+         (eth_status_linux.autoneg == AUTONEG_ENABLE);
+      status->is_autonegotiation_supported =
+         ((eth_status_linux.advertising & ADVERTISED_Autoneg) > 0);
```
Comment (fredrikdanielmoller) [inline src/ports/linux/pnal.c:382]:
```suggestion
         (eth_status_linux.advertising & ADVERTISED_Autoneg);
```

**Review by fredrikdanielmoller — CHANGES_REQUESTED** [review]

**Review by olbjo — APPROVED** [review]
Looks good to me.

I started a review yesterday which I did not have to complete until now. A bit unclear what "changes" I just approved 😃

File: src/pnal.h [hunk 2]
```diff
@@ -251,11 +270,27 @@ void pnal_eth_get_status (
...
+int pnal_get_interface_index (const char * interface_name);
+
+/**
+ * Get network interface statistics
+ *
+ * Despite the name, the statistics is for one Profinet port
+ * (not one Profinet interface).
+ *
+ * @param interface_name          In:  Ethernet interface name for example eth0
+ * @param pnal_interface_stats_t  Out: Returned statistics
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/pnal.h]: Done

File: src/ports/linux/pnal.c [hunk 3]
```diff
@@ -256,6 +255,65 @@ int pnal_set_ip_suite (
...
+   int ret = -1;
+
+   if (getifaddrs (&p_ifaddr) != 0)
+   {
+      return ret;
+   }
+
+   for (p_temp = p_ifaddr; p_temp != NULL; p_temp = p_temp->ifa_next)
+   {
+      printf ("NAME %s\n", p_temp->ifa_name);
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/ports/linux/pnal.c]: Removed
Done

File: src/device/pf_cmina.c [hunk 4]
```diff
@@ -1318,17 +1318,27 @@ void pf_ip_address_show (uint32_t ip)
...
-      "s\n",
-      net->interface_name,
-      net->interface_statistics.if_in_octets,
-      net->interface_statistics.if_in_errors,
-      net->interface_statistics.if_in_discards,
-      net->interface_statistics.if_out_octets,
-      net->interface_statistics.if_out_errors,
-      net->interface_statistics.if_out_discards);
+   pnal_interface_stats_t stats;
+
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/device/pf_cmina.c:1327]: Implemented iteration over all ports

File: src/ports/rt-kernel/pnal.c [hunk 5]
```diff
@@ -101,6 +101,40 @@ int pnal_get_ip_suite (
...
+   /* TODO Read statistics */
+   memset (interface_stats, 0, sizeof (*interface_stats));
+   return 0;
+}
+
+int pnal_get_interface_index (const char * interface_name)
+{
+   // TODO Find interface from name. Compilation error now.
+
+   struct netif * netif = interface[handle->if_id];
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/ports/rt-kernel/pnal.c]: Thanks. Did check for NULL.
Done.

File: src/ports/linux/pnal.c [hunk 6]
```diff
@@ -256,6 +257,187 @@ int pnal_set_ip_suite (
...
+   /* TODO Differentiate between copper, fiber and radio.
+           Better handling of edge cases. */
+
+   if (duplex == DUPLEX_FULL)
+   {
+      switch (speed)
+      {
+      case SPEED_10:
+         return PNAL_ETH_MAU_COPPER_10BaseT;
+         break;
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/ports/linux/pnal.c]: Done

File: src/ports/linux/pnal.c [hunk 7]
```diff
@@ -256,6 +257,187 @@ int pnal_set_ip_suite (
...
+      case SPEED_100:
+         return PNAL_ETH_MAU_COPPER_100BaseTX_FULL_DUPLEX;
+         break;
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/ports/linux/pnal.c]: Done

File: src/ports/linux/pnal.c [hunk 8]
```diff
@@ -256,6 +257,187 @@ int pnal_set_ip_suite (
...
+      case SPEED_1000:
+         return PNAL_ETH_MAU_COPPER_1000BaseT_FULL_DUPLEX;
+         break;
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/ports/linux/pnal.c]: Done

File: src/ports/linux/pnal.c [hunk 9]
```diff
@@ -256,6 +257,187 @@ int pnal_set_ip_suite (
...
+      }
+   }
+   else if (duplex == DUPLEX_HALF)
+   {
+      switch (speed)
+      {
+      case SPEED_10:
+         return PNAL_ETH_MAU_COPPER_10BaseT;
+         break;
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/ports/linux/pnal.c]: Done

File: src/ports/linux/pnal.c [hunk 10]
```diff
@@ -256,6 +257,187 @@ int pnal_set_ip_suite (
...
+      case SPEED_100:
+         return PNAL_ETH_MAU_COPPER_100BaseTX_HALF_DUPLEX;
+         break;
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/ports/linux/pnal.c]: Done

File: src/ports/linux/pnal.c [hunk 11]
```diff
@@ -256,6 +257,187 @@ int pnal_set_ip_suite (
...
+      case SPEED_1000:
+         return PNAL_ETH_MAU_COPPER_1000BaseT_HALF_DUPLEX;
+         break;
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/ports/linux/pnal.c]: Done

File: src/ports/linux/pnal.c [hunk 12]
```diff
@@ -256,6 +257,187 @@ int pnal_set_ip_suite (
...
+ * Calculate advertised capabilites
+ *
+ * @param advertised       In:    Linux advertised capabilites as a bitfield
+ * @return Profinet advertised capabilites as a bitfield
+ */
+static uint16_t calculate_capabilities (uint32_t advertised)
+{
+   uint16_t out = 0;
+
+   if ((advertised & ADVERTISED_10baseT_Half) > 0)
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/ports/linux/pnal.c]: Done

File: src/ports/linux/pnal.c [hunk 13]
```diff
@@ -256,6 +257,187 @@ int pnal_set_ip_suite (
...
+   struct ethtool_cmd eth_status_linux;
+   uint32_t speed = 0; /* Mbit/s */
+
+   control_socket = socket (PF_INET, SOCK_DGRAM, IPPROTO_IP);
+   if (control_socket < 0)
+   {
+      return ret;
+   }
+
+   strncpy (ifr.ifr_name, interface_name, sizeof (ifr.ifr_name));
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/ports/linux/pnal.c]: Done

File: src/ports/linux/pnal.c (hunk 1 above)
Comment (pyhys) [reply src/ports/linux/pnal.c]: Done

File: src/pnal.h [hunk 14]
```diff
@@ -84,6 +93,17 @@ typedef enum pnal_eth_mau
    PNAL_ETH_MAU_FIBER_1000BaseX_FULL_DUPLEX = 0x16,
 } pnal_eth_mau_t;
 
+/* See Profinet 2.4 section 5.2.28 and appendix W */
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/pnal.h]: Done

File: src/pnal.h [hunk 15]
```diff
@@ -251,11 +270,27 @@ void pnal_eth_get_status (
...
  *
  * @return  The interface index, or 0 if not available.
  */
-int pnal_get_interface_index (pnal_eth_handle_t * handle);
+int pnal_get_interface_index (const char * interface_name);
+
+/**
+ * Get network interface statistics
+ *
+ * Despite the name, the statistics is for one Profinet port
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/pnal.h]: Renamed function and data struct.
Done

File: src/device/pf_block_writer.c [hunk 16]
```diff
@@ -3990,6 +4003,15 @@ static void pf_put_pd_multiblock_port_and_statistics (
 {
```
Comment (fredrikdanielmoller) [inline src/device/pf_block_writer.c:4003]: Check if TODO-comment in function description could be removed

File: src/device/pf_cmina.c [hunk 17]
```diff
@@ -1317,19 +1317,62 @@ void pf_ip_address_show (uint32_t ip)
...
+   pf_port_iterator_t port_iterator;
+   pnal_port_stats_t stats;
+   const pnet_port_cfg_t * p_port_config = NULL;
+
+   if (pnal_get_port_statistics (net->fspm_cfg.if_cfg.main_port.if_name, &stats) == 0)
+   {
+      printf (
+         "Main interface %s    In: %" PRIu32 " bytes %" PRIu32 " errors %" PRIu32
+         " discards  Out: %" PRIu32 " bytes %" PRIu32 " errors %" PRIu32
+         " discard"
```
Comment (fredrikdanielmoller) [inline src/device/pf_cmina.c:1333]:
```suggestion
         " discards"
```

File: src/device/pf_cmina.c [hunk 18]
```diff
@@ -1317,19 +1317,62 @@ void pf_ip_address_show (uint32_t ip)
...
+   while (port != 0)
+   {
+      p_port_config = pf_lldp_get_port_config (net, port);
+
+      if (pnal_get_port_statistics (p_port_config->phy_port.if_name, &stats) == 0)
+      {
+         printf (
+            "Port %s    In: %" PRIu32 " bytes %" PRIu32 " errors %" PRIu32
+            " discards  Out: %" PRIu32 " bytes %" PRIu32 " errors %" PRIu32
+            " discard"
```
Comment (fredrikdanielmoller) [inline src/device/pf_cmina.c:1359]:
```suggestion
            " discards"
```

**Review by fredrikdanielmoller — CHANGES_REQUESTED** [review]

File: src/pnal.h [hunk 19]
```diff
@@ -251,27 +270,43 @@ void pnal_eth_get_status (
...
+int pnal_get_interface_index (const char * interface_name);
+
+/**
+ * Get network interface (port) statistics
+ *
+ * In Profinet naming a physical Ethernet interface is a "port".
+ * This corresponds to for example a Linux Ethernet interface.
+ *
+ * @param interface_name   In:    Ethernet interface name for example eth0
+ * @param port_stats  Out:   Returned statistics
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/pnal.h]: Done

File: src/ports/rt-kernel/pnal.c [hunk 20]
```diff
@@ -101,6 +101,43 @@ int pnal_get_ip_suite (
...
+}
+
+int pnal_eth_get_status (const char * interface_name, pnal_eth_status_t * status)
+{
+   /* TODO: Read current status */
+   status->is_autonegotiation_supported = true;
+   status->is_autonegotiation_enabled = true;
+   status->autonegotiation_advertised_capabilities =
+      PNAL_ETH_AUTONEG_CAP_10BaseT_HALF_DUPLEX |
+      PNAL_ETH_AUTONEG_CAP_10BaseT_FULL_DUPLEX;
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/ports/rt-kernel/pnal.c]: Done

File: src/ports/linux/pnal.c [hunk 21]
```diff
@@ -256,6 +257,180 @@ int pnal_set_ip_suite (
...
+{
+   return if_nametoindex (interface_name);
+}
+
+int pnal_get_port_statistics (
+   const char * interface_name,
+   pnal_port_stats_t * port_stats)
+{
+   struct ifaddrs *p_ifaddr, *p_temp;
+   struct rtnl_link_stats * p_statistics_linux = NULL;
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/ports/linux/pnal.c]: Done

File: src/device/pf_pdport.c [hunk 22]
```diff
@@ -442,7 +441,9 @@ int pf_pdport_read_ind (
 {
    int ret = -1;
    int loc_port_num = 0;
+   pnal_port_stats_t port_stats;
    pf_port_t * p_port_data = NULL;
+   const pnet_port_cfg_t * p_port_cfg = NULL;
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/device/pf_pdport.c]: Done

File: src/device/pf_block_writer.c (hunk 16 above)
Comment (pyhys) [reply src/device/pf_block_writer.c:4007]: I think @olbjo will study this in more detail in the work he is doing on multi ports right now.

File: src/device/pf_cmina.c (hunk 17 above)
Comment (pyhys) [reply src/device/pf_cmina.c]: Done

File: src/device/pf_cmina.c (hunk 18 above)
Comment (pyhys) [reply src/device/pf_cmina.c]: Done

File: src/device/pf_cmina.c [hunk 23]
```diff
@@ -1317,19 +1317,62 @@ void pf_ip_address_show (uint32_t ip)
...
-      net->interface_statistics.if_in_octets,
-      net->interface_statistics.if_in_errors,
-      net->interface_statistics.if_in_discards,
-      net->interface_statistics.if_out_octets,
-      net->interface_statistics.if_out_errors,
-      net->interface_statistics.if_out_discards);
+   int port;
+   pf_port_iterator_t port_iterator;
+   pnal_port_stats_t stats;
+   const pnet_port_cfg_t * p_port_config = NULL;
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/device/pf_cmina.c]: Done

File: src/ports/linux/pnal.c [hunk 24]
```diff
@@ -256,6 +257,180 @@ int pnal_set_ip_suite (
...
+   struct ethtool_cmd eth_status_linux;
+   uint32_t speed = 0; /* Mbit/s */
+
+   control_socket = socket (PF_INET, SOCK_DGRAM, IPPROTO_IP);
+   if (control_socket < 0)
+   {
+      return ret;
+   }
+
+   snprintf(ifr.ifr_name, sizeof (ifr.ifr_name), "%s", interface_name);
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/ports/linux/pnal.c]: Done

**Review by fredrikdanielmoller — APPROVED** [review]

**Pull request merged**